### PR TITLE
[numpy] add op random.rayleigh

### PR DESCRIPTION
**Pull request by Yiyan66**
## Description ##
add op random.rayleigh

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/random/np_rayleigh_op.h [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+              uniform_tensor.dptr_, outputs[0].dptr<OType>());
+        });
+      });
+    });
+  }
+}
+
+template<typename xpu, int ndim, typename DType>
+inline void ScalarRayleighReparamBackwardImpl(const OpContext& ctx,
+                                                 const std::vector<TBlob>& inputs,
```
Comment (haojin2) [inline src/operator/numpy/random/np_rayleigh_op.h]: alignment.

File: src/operator/numpy/random/np_rayleigh_op.h [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      ReduceWorkspaceSize<ndim, DType>(s, igrad.shape_, req[0], ograd.shape_);
+  Tensor<xpu, 1, char> workspace =
+      ctx.requested[0].get_space_typed<xpu, 1, char>(Shape1(workspace_size), s);
+  Reduce<red::sum, ndim, DType, op::mshadow_op::mul, op::mshadow_op::left>(
+      s, igrad, req[0], workspace, ograd, noise, noise);
+}
+
+template<typename xpu>
+void RayleighReparamBackward(const nnvm::NodeAttrs& attrs,
+                                const OpContext& ctx,
```
Comment (haojin2) [inline src/operator/numpy/random/np_rayleigh_op.h]: alignment

File: tests/python/unittest/test_numpy_op.py [hunk 3]
```diff
@@ -3624,6 +3624,44 @@ def _test_random_beta_range(output):
         assert _test_random_beta_range(mx_out_imperative.asnumpy()) == True
 
 
+@with_seed()
+@use_np
+def test_np_rayleigh():
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: `test_np_random_rayleigh`

File: python/mxnet/ndarray/numpy/random.py [hunk 4]
```diff
@@ -279,6 +280,44 @@ def multinomial(n, pvals, size=None):
...
 
+def rayleigh(scale=1.0, size=None, ctx=None, out=None):
+    r"""Draw samples from a Rayleigh distribution.
+
+    The :math:`\chi` and Weibull distributions are generalizations of the
+    Rayleigh.
+
+    Parameters
+    ----------
+    scale : float or array_like of floats, optional
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/random.py]: not `array_like`, we should only support either `mx.numpy.ndarray` or a float scalar.

File: python/mxnet/symbol/numpy/random.py [hunk 5]
```diff
@@ -382,6 +383,42 @@ def gamma(shape, scale=1.0, size=None, dtype=None, ctx=None, out=None):
...
 
+def rayleigh(scale=0.0, size=None, ctx=None, out=None):
+    r"""Draw samples from a Rayleigh distribution.
+
+    The :math:`\chi` and Weibull distributions are generalizations of the
+    Rayleigh.
+
+    Parameters
+    ----------
+    scale : float or _Symbol of floats
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/random.py]: not `float or _Symbol of floats`, should simply be `float or _Symbol`

File: python/mxnet/ndarray/numpy/random.py [hunk 6]
```diff
@@ -24,6 +24,7 @@
 
 
 __all__ = ['randint', 'uniform', 'normal', "choice", "rand", "multinomial", "multivariate_normal",
+           "rayleigh",
```
Comment (xidulu) [inline python/mxnet/ndarray/numpy/random.py]: Don't start a new line for only **one** item.

File: python/mxnet/numpy/random.py [hunk 7]
```diff
@@ -22,6 +22,7 @@
 
 
 __all__ = ["randint", "uniform", "normal", "choice", "rand", "multinomial", "multivariate_normal",
+           "rayleigh",
```
Comment (xidulu) [inline python/mxnet/numpy/random.py]: Don't start a new line for only one item.

File: src/operator/numpy/random/np_rayleigh_op.h [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+template <typename IType>
+struct check_legal_scale_kernel {
+  MSHADOW_XINLINE static void Map(index_t i, IType *scalar, float* flag) {
+    if (scalar[i] < 0.0) {
+      flag[0] = -1.0;
+    }
+  }
+};
+
```
Comment (haojin2) [inline src/operator/numpy/random/np_rayleigh_op.h]: 1 blank line between functions in c++

**Pull request merged**